flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -199,7 +252,32 @@ private void flush() throws InterruptedException {
      */
     private void completeRequest(Collection<RequestEntryT> failedRequestEntries) {
         inFlightRequestsCount--;
-        failedRequestEntries.forEach(bufferedRequestEntries::addFirst);
+        failedRequestEntries.forEach(failedEntry -> addEntryToBuffer(failedEntry, true));
+    }
+
+    private void addEntryToBuffer(RequestEntryT entry, boolean insertAtHead) {
+        if (bufferedRequestEntries.isEmpty() && !existsActiveTimerCallback) {
+            registerCallback();
+        }
+        if (insertAtHead) {
+            bufferedRequestEntries.addFirst(entry);
+        } else {
+            bufferedRequestEntries.add(entry);
+        }
+
+        double requestEntrySizeMB = getSizeInMB(entry);
+
+        if (insertAtHead) {
+            bufferedRequestEntriesSizeMB.addFirst(requestEntrySizeMB);
+        } else {
+            bufferedRequestEntriesSizeMB.add(requestEntrySizeMB);
+        }
+
+        bufferedRequestEntriesTotalSizeMB += requestEntrySizeMB;
+    }
```

Comment:
You are maintaining 2 queues and keeping them in sync. This is a complex model and prone to error with items becoming unsynchronised while evolving the implementation. Instead of  having 2x `Deque` I suggest you implement a wrapper than contains the record and size, then use a single queue, for example:

```
public class RequestEntryWrapper<T> {
    private final T requestEntry;
    private final long size;

    ...
}
```

Then just have:

```
private Deque<RequestEntryWrapper> theSingleQueue;
```

Reply:
It's a lot neater now with just the 1. Thanks.